### PR TITLE
Pass `allJars` when creating ScalaInstance instead of `otherJars`.

### LINKED ISSUE #3560: `scalaInstance.allJars` doesn't include scala-compiler.jar depending on Scala version
Using any `scalaVersion` setting other than `"2.12.3"`, `scalaInstance.allJars` will be missing the `scala-compiler.jar`.

IntelliJ depends on this task to setup the compiler classpath when not using sbt directly.
Corresponding ticket: https://youtrack.jetbrains.com/issue/SCL-12591

## steps

* in `build.properties`: 
```
sbt.version = 1.0.2
```

* in `build.sbt`
```
scalaVersion := "2.12.1"
```

Run `consoleProject` on sbt shell:
```
> consoleProject

scala> scalaInstance.eval.allJars
res0: Array[java.io.File] = Array(/Users/jast/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.12.1.jar, /Users/jast/.ivy2/cache/org.scala-lang.modules/scala-xml_2.12/bundles/scala-xml_2.12-1.0.6.jar, /Users/jast/.ivy2/cache/jline/jline/jars/jline-2.14.1.jar)
```

Compare using `scalaVersion := "2.12.3"`

```
> consoleProject

scala> scalaInstance.eval.allJars
res2: Array[java.io.File] = Array(/Users/jast/.sbt/boot/scala-2.12.3/lib/jline.jar, /Users/jast/.sbt/boot/scala-2.12.3/lib/scala-compiler.jar, /Users/jast/.sbt/boot/scala-2.12.3/lib/scala-library.jar, /Users/jast/.sbt/boot/scala-2.12.3/lib/scala-reflect.jar, /Users/jast/.sbt/boot/scala-2.12.3/lib/scala-xml_2.12.jar)    
```

## notes

sbt version: 1.0.0 - 1.0.2

### PR DESCRIPTION
**Pull request by jastice**
Fixes #3560

### CODE REVIEW

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3561/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/185/console">is in progress...</a>

**Review by eed3si9n — APPROVED** [review]

**Comment by typesafe-tools** [comment]
[//]: # (Refs: pull/3561/head-1.x-1.x-1.x-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/185/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/3561">pull/3561/head      </A> | sbt/sbt@45d69f37624193541837a169d7780dc7add96329
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/1.x">1.x                 </A> | sbt/zinc@ef3a2c48a4e6b6126c8f1d0a99e530d64925c7d3
  io                | <A HREF="https://github.com/sbt/io/tree/1.x">1.x                 </A> | sbt/io@62004b28dfcdc33273ebb07a563e78178f6eb21d
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.x">1.x                 </A> | sbt/librarymanagement@04193716da1d7454d1d7c5935b396ebeefc8c2d4
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | sbt/util@cba6f83447b84cc1e967f25a930d4f411f6f0664
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/3561/head&zinc=1.x&io=1.x&librarymanagement=1.x&util=1.x&website=1.x&deploy_to_bintray=false">(restart)</A>

**Comment by dwijnand** [comment]
Before merging this I would've liked to understand what changes caused the regression. If you look at sbt 0.13.16 then "otherJars" was passed to ScalaInstance, along with libraryJar and compilerJar: https://github.com/sbt/sbt/blob/v0.13.16/main/src/main/scala/sbt/Defaults.scala#L443

We run the risk of causing regressions like this..

**Comment by jastice** [comment]
In 0.13.x, `allJars` was constructed from the parameters:
```
  def allJars: Array[File] = jars.toArray
  def jars = libraryJar :: compilerJar :: extraJars.toList
```

In 1.0, `allJars` is a parameter and `otherJars` is derived:
```
def otherJars = (allJars.toSet - libraryJar - compilerJar).toArray
```

**Comment by dwijnand** [comment]
Looks like the original change was in https://github.com/sbt/zinc/pull/5/commits/ecf121a7ae7ba99cddad69c91117948174dbfc15#diff-28f65fcd38afbb29ef614d184684ac2c.

**Comment by eed3si9n** [comment]
When I reviewed the pr I did go through the history of ScalaInstance and found the same thing. I guess I should've reported my findings here.

File: main/src/main/scala/sbt/Defaults.scala [hunk 1]
```diff
@@ -601,19 +601,18 @@ object Defaults extends BuildCommon {
...
     val libraryJar = file(ScalaArtifacts.LibraryID)
     val binVersion = scalaBinaryVersion.value
     val compilerJar =
       if (ScalaInstance.isDotty(scalaVersion.value))
         file(ScalaArtifacts.dottyID(binVersion))
       else file(ScalaArtifacts.CompilerID)
-    val otherJars = allFiles.filterNot(x => x == libraryJar || x == compilerJar)
     new ScalaInstance(scalaVersion.value,
-                      makeClassLoader(state.value)(libraryJar :: compilerJar :: otherJars.toList),
+                      makeClassLoader(state.value)(allJars.toList),
```
Comment (fommil) [inline main/src/main/scala/sbt/Defaults.scala:612]: interesting, but doesn't this mean the scalac jar isn't on the classpath?

I once used a monkey patch to allow custom classes to go ahead of the scala-compiler jar https://gitlab.com/fommil/attic/blob/master/sbt-scala-monkey/src/main/scala/ScalaMonkeyPlugin.scala#L60

Comment (fommil) [reply main/src/main/scala/sbt/Defaults.scala:612]: oh, I stand corrected... I was also filtering out the compiler / library jars.

**Pull request merged**